server/server.go

```diff
@@ -433,3 +435,50 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
 
 	return func() { watcher.Close() }, watcherChan, nil
 }
+
+// writeHTTPError writes a JSON error response
+func writeHTTPError(w http.ResponseWriter, r *http.Request, statusCode int, errorCode, errorDescription string) {
+	args := []any{
+		slog.Int("status", statusCode),
+		slog.String("method", r.Method),
+		slog.String("path", r.URL.Path),
+		slog.String("error", errorCode),
+		slog.String("description", errorDescription),
+	}
+
+	switch statusCode {
+	case http.StatusForbidden, http.StatusUnauthorized:
+		slog.Warn("HTTP error", args...)
+	case http.StatusInternalServerError:
+		slog.Error("HTTP error", args...)
+	default:
+		slog.Debug("HTTP error", args...)
+	}
+
+	w.Header().Set("Cache-Control", "no-store")
+	w.Header().Set("Pragma", "no-cache")
+	w.WriteHeader(statusCode)
+
+	acceptHeader := r.Header.Get("Accept")
+	switch {
+	case strings.Contains(acceptHeader, "text/html"):
+		fallthrough
+	case strings.Contains(acceptHeader, "text/plain"):
+		w.Header().Set("Content-Type", "text/plain; charset=UTF-8")
+		fmt.Fprintf(w, "Error %d: %s - %s", statusCode, errorCode, errorDescription)
+	default:
```

Comment:
seems odd to send back a JSON if the client didn't specify `Accept`.
should plaintext be the default, and JSON only when `Accept: application/json`?

Reply:
Picking JSON is a better default since most requests to tsidp endpoints will be an Oauth client of some sort instead of a user.